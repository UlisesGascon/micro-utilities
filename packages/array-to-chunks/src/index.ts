export function chunkArray (arr:any[], chunkSize:number):any[] { 
  if(!Array.isArray(arr)) throw new Error('arr must be an array')
  if(!Number.isInteger(chunkSize) && chunkSize >= 0) throw new Error('chunkSize must be a positive integer')

  const res: any[] = []
  for (let i = 0; i < arr.length; i += chunkSize) {
      const chunk = arr.slice(i, i + chunkSize);
      res.push(chunk);
  }
  return res
}