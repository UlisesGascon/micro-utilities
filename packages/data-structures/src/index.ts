class Node {
  value: any;
  next: Node | null;

  constructor(value: any){
      this.value = value;
      this.next = null;
  }
}

export class LinkedList {
  head: Node | null;
  tail: Node | null;
  length: number;

  constructor (value:any) {
    const node = new Node(value)
    this.head = node
    this.tail = this.head
    this.length = 1
  }

  public push (value: any):LinkedList {
    const node = new Node(value)
    if(this.tail != null) {
      this.tail.next = node
    }
    this.tail = node
    this.length++
    return this
  }

  public pop () :any | undefined {
    if(this.head == null) {
      return undefined
    }

    let temp = this.head

    if(this.length === 1) {
      this.head = null
      this.tail = null
      this.length = 0
      return temp.value
    }

    while (temp.next != null) {
      temp = temp.next
    }

    this.tail = temp
    this.tail.next = null
    this.length--
    return temp.value
  }

  public shift (value:any) :any {
    if(this.head == null) {
      return undefined
    }

    const temp = this.head
    this.head = temp.next
    this.length--
    if(this.length === 0) {
      this.tail = null
    }
    return temp.value
  }

  public unshift (value:any) :LinkedList {
    const node = new Node(value)
    if(this.head == null) {
      this.head = node
      this.tail = this.head
    } else {
      node.next = this.head
      this.head = node
    }
    this.length++
    return this
  }

  public purge () :LinkedList {
    this.head = null
    this.tail = null
    this.length = 0
    return this
  }
}
