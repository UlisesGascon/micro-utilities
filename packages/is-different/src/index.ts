export function isDifferent(firstValue: any, secondValue: any): boolean {
  return JSON.stringify(firstValue) !== JSON.stringify(secondValue);
}
