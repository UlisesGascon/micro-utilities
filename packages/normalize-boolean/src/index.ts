export function normalizeBoolean (value:any):boolean {
    // Ignore non-primitive values
    if (!['boolean', 'string', 'number'].includes(typeof value)) {
      return false
    }
  
    const normalizedValue = !isNaN(parseInt(value)) ? parseInt(value) : value
  
    // Add support for string values
    if (typeof normalizedValue === 'string') {
      return ['true', 'on', 'yes', 'y'].includes(normalizedValue.toLowerCase().trim())
    }
  
    // Add support for number values
    if (typeof normalizedValue === 'number') {
      return normalizedValue === 1
    }
  
    return Boolean(value)
}