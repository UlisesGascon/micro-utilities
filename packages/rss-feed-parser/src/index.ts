import { XMLParser, XMLValidator} from 'fast-xml-parser'
import { type RssParseReturn } from './types';

const parser = new XMLParser();

export function rssParse(rssContent: string): RssParseReturn {
  const result:any = XMLValidator.validate(rssContent);
  if (result.err !== undefined) {
    throw new Error('XML is invalid');
  }
  
  const jObj = parser.parse(rssContent);

  if (jObj?.rss?.channel === undefined) {
    throw new Error('XML has not valid rss format');
  }

  const {item: items, ...metadata} = jObj.rss.channel;

  return {
    metadata,
    items: items ?? []
  }
}
