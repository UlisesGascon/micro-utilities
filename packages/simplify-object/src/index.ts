interface simplifyPayloadType {
  include?: string[]
  exclude?: string[] 
}

type OpenObjectType = Record<string, any>;

const isObject = (obj:object):boolean => {
  return typeof obj === 'object' && !Array.isArray(obj) && obj !== null
}

export function simplifyObject (baseObj:OpenObjectType, operation:simplifyPayloadType):OpenObjectType {
  if(!isObject(baseObj)) throw new Error('baseObj must be an object')
  if(!isObject(operation)) throw new Error('operation must be an object')

  if(Array.isArray(operation.include) && Array.isArray(operation.exclude)) throw new Error('operation must have either include or exclude')
  if(!Array.isArray(operation.include) && !Array.isArray(operation.exclude)) throw new Error('operation must have either include or exclude')

  if(Array.isArray(operation.include) && (operation.include.length === 0)) throw new Error('operation.include must be an array with properties')
  if(Array.isArray(operation.exclude) && (operation.exclude.length === 0)) throw new Error('operation.exclude must be an array with properties')
  
  const { include=[], exclude=[] } = operation 

  const simplifiedObject:OpenObjectType = {}

  Object.keys(baseObj).forEach(key => {
      if(((include.length > 0) && include.includes(key)) || (exclude.length > 0) && !exclude.includes(key)) {
        simplifiedObject[key] = baseObj[key]
      }
  })
  
  return simplifiedObject
}