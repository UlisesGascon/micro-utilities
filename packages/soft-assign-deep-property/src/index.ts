type OpenObjectType = Record<string, any>;

const isObject = (obj: object): boolean => {
  return typeof obj === "object" && !Array.isArray(obj) && obj !== null;
};

export function softAssign(
  obj: OpenObjectType,
  keys: string[],
  value: any
): void {
  if (!isObject(obj)) throw new Error("obj must be an object");
  if (!Array.isArray(keys) || keys.length === 0)
    throw new Error("keys must be an Array with at least one element");

  const lastKeyIndex = keys.length - 1;
  for (let i = 0; i < lastKeyIndex; ++i) {
    const key = keys[i];
    if (!(key in obj)) {
      obj[key] = {};
    }
    obj = obj[key];
  }
  obj[keys[lastKeyIndex]] =
    obj[keys[lastKeyIndex]] !== undefined ? obj[keys[lastKeyIndex]] : value;
}
