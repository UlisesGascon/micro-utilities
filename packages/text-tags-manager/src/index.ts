interface Settings {
  original: string;
  replacementSegment: string;
  startTag: string;
  endTag: string;
}

export function updateOrCreateSegment (settings:Settings):string {
  const { original, replacementSegment, startTag, endTag } = settings

  if([original, replacementSegment, startTag, endTag].some(prop => typeof prop !== 'string')){
    throw new Error('settings must be an object with the following properties as string values: original, replacementSegment, startTag, endTag')
  }

  const startTagIndex = original.indexOf(startTag)
  const endTagIndex = original.indexOf(endTag)


  // Throw an error if only one tag is found
  if((startTagIndex === -1 && endTagIndex > -1) || (endTagIndex === -1 && startTagIndex > -1)) {
    throw new Error(`Only one tag found in the content, expected to find both '${startTag}' and '${endTag}'`)
  }

  // If the start tag is found after the end tag then throw an error
  if(startTagIndex > endTagIndex){
    throw new Error(`The tags are not in the correct order, expected to find '${startTag}' before '${endTag}'`)
  }

  // If no tags found then return the replacement segment wrapped in the tags
  if(startTagIndex === -1 && endTagIndex === -1) {
    return `${startTag}\n${replacementSegment}\n${endTag}`
  }

    
  // If tags found then replace the content between the tags with the replacement segment
  const contentBefore = original.slice(0, startTagIndex)
  const contentAfter = original.slice(endTagIndex + endTag.length)
  return `${contentBefore}${startTag}\n${replacementSegment}\n${endTag}${contentAfter}`
}