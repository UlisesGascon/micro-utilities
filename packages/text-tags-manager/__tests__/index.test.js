import { updateOrCreateSegment } from '../src'

describe('text-tags-manager', () => {
  let textWithTags = null;
  let settings = null;
  const startTag = '<!-- MY-TAG:START -->'
  const endTag = '<!-- MY-TAG:END -->'
  const replacementSegment = 'This is my new text, you know... the same but more cool!'


  beforeEach(() => {
    textWithTags = `
    This text should remind the same...

<!-- MY-TAG:START -->
The old text
<!-- MY-TAG:END -->

This text will be also unaffected`
    settings = {
      startTag,
      endTag,
      replacementSegment,
      original: textWithTags
    }
  })

  it('First argument must be an object with all the properties', () => {
    const errorMessage = 'settings must be an object with the following properties as string values: original, replacementSegment, startTag, endTag'
    expect(() => updateOrCreateSegment({...settings, startTag: undefined})).toThrow(errorMessage)
    expect(() => updateOrCreateSegment({...settings, endTag: undefined})).toThrow(errorMessage)
    expect(() => updateOrCreateSegment({...settings, replacementSegment: undefined})).toThrow(errorMessage)
    expect(() => updateOrCreateSegment({...settings, original: undefined})).toThrow(errorMessage)
  })

  it('Should throw an error if only one tag is included', () => {
    const errorMessage = `Only one tag found in the content, expected to find both '${startTag}' and '${endTag}'`
    expect(() => updateOrCreateSegment({
      ...settings,
      original: textWithTags.replace(endTag, '')
    })).toThrow(errorMessage)
  })

  it('Should throw an error if the tags are not in the correct order', () => {
    const errorMessage = `The tags are not in the correct order, expected to find '${startTag}' before '${endTag}'`
    expect(() => updateOrCreateSegment({
      ...settings,
      original: textWithTags.replace(endTag, startTag).replace(startTag, endTag)
    })).toThrow(errorMessage)
  })

  it('If no tags found then return the replacement segment wrapped in the tags', () => {
    const textWithoutTags = 'This is a text without tags'
    const result = updateOrCreateSegment({
      ...settings,
      original: textWithoutTags
    })
    expect(result).toMatchSnapshot()
  })

  it('If tags found then replace the content between the tags with the replacement segment', () => {
    const result = updateOrCreateSegment(settings)
    expect(result).toMatchSnapshot()
  })

})
