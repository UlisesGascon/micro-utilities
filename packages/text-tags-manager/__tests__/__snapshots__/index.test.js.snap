// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text-tags-manager If no tags found then return the replacement segment wrapped in the tags 1`] = `
"<!-- MY-TAG:START -->
This is my new text, you know... the same but more cool!
<!-- MY-TAG:END -->"
`;

exports[`text-tags-manager If tags found then replace the content between the tags with the replacement segment 1`] = `
"
    This text should remind the same...

<!-- MY-TAG:START -->
This is my new text, you know... the same but more cool!
<!-- MY-TAG:END -->

This text will be also unaffected"
`;
